internal/server/iito.go

```diff
@@ -48,56 +48,61 @@ func (s *SOMASServer) runGiftSession() error {
 	return nil
 }
 
-func (s *SOMASServer) getGiftRequests() shared.GiftDict {
-	giftRequestDict := shared.GiftDict{}
+// GetGiftRequests collects all gift requests from the clients in a map
+func (s *SOMASServer) getGiftRequests() map[shared.ClientID]shared.GiftRequestDict {
+	giftRequestDict := map[shared.ClientID]shared.GiftRequestDict{}
 	for id, client := range s.clientMap {
-		giftRequestDict[id] = client.RequestGift()
+		giftRequestDict[id] = client.GetGiftRequests()
 	}
 	return giftRequestDict
 }
-func (s *SOMASServer) getGiftOffers(giftRequestDict shared.GiftDict) (map[shared.ClientID]shared.GiftDict, error) {
-	giftOfferDict := map[shared.ClientID]shared.GiftDict{}
-	var err error
+
+// GetGiftOffers collects all responses from clients to their requests in a map
+func (s *SOMASServer) getGiftOffers(totalRequests map[shared.ClientID]shared.GiftRequestDict) (map[shared.ClientID]shared.GiftOfferDict, error) {
+	totalOffers := map[shared.ClientID]shared.GiftOfferDict{}
+	// Loop over each team
 	for id, client := range s.clientMap {
-		giftOfferDict[id], err = client.OfferGifts(giftRequestDict)
+		var offer map[shared.ClientID]shared.GiftOffer
+		var err error
+		// Gather all the requests made to this team
+		requestsToThisTeam := shared.GiftRequestDict{}
+		for fromTeam, indivRequests := range totalRequests {
+			requestsToThisTeam[fromTeam] = indivRequests[id]
```

Comment:
This will break if indivRequests does not have a value for the corresponding id.
I.e. if team 1 does not request anything from team 2, when team 2's requests are being packaged it will try to find team 1's request for team 2 and then break

A similar issues is found on line 84 and 99

Reply:
As @QFSW and I discussed, we should make data types to represent this outcome, and let the server do a "validation" of sorts to make sure data is in the right shape at each phase. This also applies to fixing the other issue you mentioned of clients mutating data inappropriately.

The reason everything is done in atomic map operations in this first draft is because we want clients to be able to make all their decisions at once, instead of being asked for a response on an individual basis: A client who does not like another may still accept a gift from it if it is the only gift it received and it needs resources.